core/src/main/scala/org/apache/spark/Dependency.scala

```diff
@@ -85,14 +84,28 @@ class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
   private[spark] val combinerClassName: Option[String] =
     Option(reflect.classTag[C]).map(_.runtimeClass.getName)
 
-  val shuffleId: Int = _rdd.context.newShuffleId()
-
   val shuffleHandle: ShuffleHandle = _rdd.context.env.shuffleManager.registerShuffle(
     shuffleId, _rdd.partitions.size, this)
+}
+
+
+private[spark] abstract class BaseShuffleDependency[T: ClassTag](
```

Comment:
The idea behind this change is that our existing `ShuffleDependency` is a poor fit for non-record-oriented shuffle or other low-level shuffle APIs, but we can't / shouldn't break or change it since it's a `@DeveloperApi` that several people depend on.

Changing the inheritance hierarchy here should not create any binary compatibility problems.